static/js/list_widget.js

```diff
@@ -354,6 +352,54 @@ export function create($container, list, opts) {
         }
     };
 
+    widget.increase_rendered_offset = function () {
+        meta.offset = Math.min(meta.offset + 1, meta.filtered_list.length);
+    };
+
+    widget.reduce_rendered_offset = function () {
+        meta.offset = Math.max(meta.offset - 1, 0);
+    };
+
+    widget.remove_rendered_row = function (rendered_row) {
+        rendered_row.remove();
+        // We removed a rendered row, so we need to reduce one offset.
+        widget.reduce_rendered_offset();
+    };
+
+    widget.insert_rendered_row = function (item) {
+        if (meta.filtered_list.length <= 2) {
+            // Avoids edge cases for us and could be faster too.
+            widget.clean_redraw();
+        }
+        if (!opts.filter.predicate(item)) {
+            return;
+        }
+        // We need to insert the row for it to be displayed at the
+        // correct position. filtered_list must contain the new item
+        // since we know it is not hidden from the above check.
```

Comment:
Just to make sure I understand the logic here -- is the model that we have a new item that we've added to the data set (and thus we know it's already in `meta.filtered_list`), and `opts.filter.predicate(item)` has confirmed that it should be displayed?

Do we need to add a comment at the top of this function explaining it has the precondition that you know the item in question to the `list_widget`'s data object before calling this?

(I assume the caller of `inplace_rerender` is responsible for doing that in Recent Conversations in particular?)

Reply:
> Do we need to add a comment at the top of this function explaining it has the precondition that you know the item in question to the list_widget's data object before calling this?

Yeah, makes sense.